openstack_hypervisor/fileserver/server.py

```diff
@@ -0,0 +1,400 @@
+    except Exception:
+        body = request.body_file.read()
+    if not body:
+        return {}
+    try:
+        return json.loads(body.decode("utf-8"))
+    except Exception as exc:
+        raise BadRequestError(f"invalid JSON body: {exc}")
+
+
+def _ok(payload: Dict[str, Any]) -> Response:
+    """Return a JSON 200 OK response with the provided payload."""
+    return Response(json_body=payload)
+
+
+def _error(status: int, detail: str) -> Response:
+    """Return a JSON error response with the given HTTP status and detail."""
+    return Response(json_body={"detail": detail}, status=status)
+
+
+def create_file_ep(request: Request) -> Response:
+    """Create an empty file at the requested destination path."""
+    try:
+        payload = _json(request)
+        dst_path = payload.get("dst_path")
+        if not dst_path:
+            raise BadRequestError("dst_path is required")
+        dst = Path(dst_path)
+        dst.parent.mkdir(parents=True, exist_ok=True)
+        try:
+            with open(dst, "xb"):
+                pass
+        except FileExistsError:
+            pass
+        return _ok({"created": True, "path": str(dst)})
+    except BadRequestError as exc:
+        return _error(400, str(exc))
+    except Exception as exc:
+        LOG.exception("create_file failed: %s", exc)
+        return _error(500, str(exc))
+
+
+def remove_file_ep(request: Request) -> Response:
+    """Remove a file at the given destination path (if present)."""
+    dst_path = request.GET.get("dst_path")
+    if not dst_path:
+        return _error(400, "dst_path is required")
+    dst = Path(dst_path)
+    try:
+        Path(dst).unlink(missing_ok=True)
+        return _ok({"removed": True, "path": str(dst)})
+    except Exception as exc:
+        LOG.exception("remove_file failed: %s", exc)
+        return _error(500, str(exc))
+
+
+def create_dir_ep(request: Request) -> Response:
+    """Create a directory (parents as needed) at the destination path."""
+    try:
+        payload = _json(request)
+        dst_path = payload.get("dst_path")
+        if not dst_path:
+            raise BadRequestError("dst_path is required")
+        dst = Path(dst_path)
+        Path(dst).mkdir(parents=True, exist_ok=True)
+        return _ok({"created": True, "path": str(dst)})
+    except BadRequestError as exc:
+        return _error(400, str(exc))
+    except Exception as exc:
+        LOG.exception("create_dir failed: %s", exc)
+        return _error(500, str(exc))
+
+
+def remove_dir_ep(request: Request) -> Response:
+    """Remove an empty directory at the destination path."""
+    dst_path = request.GET.get("dst_path")
+    if not dst_path:
+        return _error(400, "dst_path is required")
+    dst = Path(dst_path)
+    try:
+        Path(dst).rmdir()
+        return _ok({"removed": True, "path": str(dst)})
+    except Exception as exc:
+        LOG.exception("remove_dir failed: %s", exc)
+        return _error(500, str(exc))
```

Comment:
if someone attempts to remove a non-empty directory, the error should be 4xx, not 500, since it's the user's request that was incorrect when requested to remove a non-empty directory